src/Language/DifferentialDatalog/Type.hs

```diff
@@ -275,6 +274,16 @@ exprNodeType' d _   (EField p (Just e) f) = do
                                  return $ fieldType fld
          _           -> err (pos e) $ "Expression is not a struct"
 
+
+exprNodeType' _ ctx (ETupField p Nothing _)  = eunknown p ctx
+exprNodeType' d _   (ETupField p (Just e) i) = do
+    case typ' d e of
+         t@TTuple{} ->
+             if ((toInteger $ length $ typeTupArgs t) <= i) || (0 > i)
```

Comment:
You should be able to use `check` instead of if-then-else here. It's shorter and more idiomatic